processing/src/main/java/org/apache/druid/query/union/UnionQueryRunner.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.union;
+
+import org.apache.druid.java.util.common.guava.Sequence;
+import org.apache.druid.java.util.common.guava.Sequences;
+import org.apache.druid.query.Query;
+import org.apache.druid.query.QueryLogic;
+import org.apache.druid.query.QueryPlus;
+import org.apache.druid.query.QueryRunner;
+import org.apache.druid.query.QueryRunnerFactoryConglomerate;
+import org.apache.druid.query.QuerySegmentWalker;
+import org.apache.druid.query.context.ResponseContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+class UnionQueryRunner implements QueryRunner<Object>
+{
+  private final QueryRunnerFactoryConglomerate conglomerate;
+  private final QuerySegmentWalker walker;
+  private final List<QueryRunner> runners;
+
+  public UnionQueryRunner(
+      UnionQuery query,
+      QueryRunnerFactoryConglomerate conglomerate,
+      QuerySegmentWalker walker)
+  {
+    this.conglomerate = conglomerate;
+    this.walker = walker;
+    this.runners = makeSubQueryRunners(query);
+  }
+
+  private List<QueryRunner> makeSubQueryRunners(UnionQuery unionQuery)
+  {
+    List<QueryRunner> runners = new ArrayList<>();
+    for (Query<?> query : unionQuery.queries) {
+      runners.add(buildRunnerFor(query));
+    }
+    return runners;
+  }
+
+  private QueryRunner<?> buildRunnerFor(Query<?> query)
+  {
+    QueryLogic queryLogic = getQueryLogicFor(query);
+    return queryLogic.entryPoint(query, walker);
+  }
+
+  private QueryLogic getQueryLogicFor(Query<?> query)
+  {
+    QueryLogic queryLogic = conglomerate.getQueryLogic(query);
+    if (queryLogic != null) {
+      return queryLogic;
+    }
+    return new ToolChestBasedQueryLogic(conglomerate.getToolChest(query));
```

Comment:
If I'm reading this code correctly, you have this object basically in order to wrap the QueryRunner such that it produces results with the correct output for the Union query.  That's an absolutely necessary bit of logic.  I don't think it's really correct to bundle that logic up into a `QueryLogic`.  A `QueryLogic` should be something that contains the nuggets of query processing and should be associated with the query type itself.  A developer would choose to implement a QueryLogic, but the existence of the `ToolChestBasedQueryLogic` makes it seem as if the class is a compatibility layer that can provide a quick "lift-and-shift" of existing `QueryToolChest` objects into the `QueryLogic` interface.  But, I don't think the implementation is actually that robust, it looks to be more specific to what this `UnionQueryRunner` needs.

So, I suggest that you inline that logic here.  Specifically, you are returning this object and then immediately calling `.entryPoint()` on it.  In that case, you should be able to just put all of that code here and wrap the `QueryRunner` that is returned from `query.getRunner(walker)`.